### ISSUE TITLE
Implicit conversion from T[] to Span<T> generates poor code

### ISSUE THREAD

**Issue by mikedn** [dotnet/coreclr#15782]
```C#
static Span<int> Test(int[] a) => a;
```
generates:
```asm
G_M33790_IG01:
G_M33790_IG02:
       4885D2               test     rdx, rdx
       7507                 jne      SHORT G_M33790_IG03
       33C0                 xor      rax, rax
       4533C0               xor      r8d, r8d
       EB18                 jmp      SHORT G_M33790_IG04
G_M33790_IG03:
       488D4210             lea      rax, bword ptr [rdx+16]
       4C8BC0               mov      r8, rax  ; meh
       498BC0               mov      rax, r8
       4C8BC0               mov      r8, rax
       8B4208               mov      eax, dword ptr [rdx+8]
       8BD0                 mov      edx, eax  ; meh again
       498BC0               mov      rax, r8
       448BC2               mov      r8d, edx
G_M33790_IG04:
       488901               mov      bword ptr [rcx], rax
       44894108             mov      dword ptr [rcx+8], r8d
       488BC1               mov      rax, rcx
G_M33790_IG05:
       C3                   ret
```

category:cq
theme:register-allocator
skill-level:expert
cost:medium

**Comment by AndyAyersMS** [comment]
Reminiscent of #11390 (though I've thought that before and been wrong). Which constructor gets invoked?

**Comment by mikedn** [comment]
The implicit conversion does `return a == null ? default : new Span<int>(a);`.

Need to take a closer look at #11390 to see if it's the same thing. But in any case, those register copies are a symptom of copy propagation being hit by a truckload of fail.

**Comment by mikedn** [comment]
`static Span<int> Test(int[] a) => new Span<int>(a);` generates the expected code:
```asm
G_M33789_IG01:
       4883EC28             sub      rsp, 40
G_M33789_IG02:
       4885D2               test     rdx, rdx
       7415                 je       SHORT G_M33789_IG05
G_M33789_IG03:
       488D4210             lea      rax, bword ptr [rdx+16]
       8B5208               mov      edx, dword ptr [rdx+8]
       488901               mov      bword ptr [rcx], rax
       895108               mov      dword ptr [rcx+8], edx
       488BC1               mov      rax, rcx
G_M33789_IG04:
       4883C428             add      rsp, 40
       C3                   ret
G_M33789_IG05:
       B902000000           mov      ecx, 2
       E8C854A45E           call     System.ThrowHelper:ThrowArgumentNullException(int)
       CC                   int3
```

**Comment by mikedn** [comment]
The reason copies are introduced in the first place may be different in 11390. But the reason they are not removed is the same in both 11390 and this issue - copy propagation.

My example generates the following IR;
```
***** BB04, stmt 6
     (  9,  8) [000130] ------------              *  STMT      void  (IL 0x000...  ???)
N004 (  1,  1) [000158] ------------              |        /--*  CNS_INT   long   16 field offset Fseq[m_arrayData] $280
N005 (  3,  3) [000159] ------------              |     /--*  ADD       byref  $2c0
N003 (  1,  1) [000157] ------------              |     |  \--*  LCL_VAR   ref    V01 arg0         u:2 $c0
N006 (  5,  5) [000160] ---XG--N----              |  /--*  COMMA     byref  $2c1
N002 (  2,  2) [000156] ---X---N----              |  |  \--*  NULLCHECK byte   $241
N001 (  1,  1) [000155] ------------              |  |     \--*  LCL_VAR   ref    V01 arg0         u:2 $c0
N008 (  9,  8) [000129] -A-XG---R---              \--*  ASG       byref  $2c1
N007 (  3,  2) [000128] D------N----                 \--*  LCL_VAR   byref  V08 tmp6         d:2 $2c0

***** BB04, stmt 7
     (  7,  5) [000102] ------------              *  STMT      void  (IL 0x000...  ???)
N001 (  3,  2) [000127] ------------              |  /--*  LCL_VAR   byref  V08 tmp6         u:2 (last use) $2c0
N003 (  7,  5) [000100] -A------R---              \--*  ASG       byref  $2c0
N002 (  3,  2) [000099] D------N----                 \--*  LCL_VAR   byref  V15 tmp13        d:2 $2c0

***** BB04, stmt 8
     (  7,  5) [000107] ------------              *  STMT      void  (IL 0x000...  ???)
N001 (  3,  2) [000166] -------N----              |  /--*  LCL_VAR   byref  V15 tmp13        u:2 (last use) $2c0
N003 (  7,  5) [000167] -A------R---              \--*  ASG       byref  $2c0
N002 (  3,  2) [000165] D------N----                 \--*  LCL_VAR   byref  V11 tmp9         d:2 $2c0

***** BB04, stmt 9
     (  7,  6) [000115] ------------              *  STMT      void  (IL 0x000...  ???)
N002 (  3,  3) [000111] ---X--------              |  /--*  ARR_LENGTH int    $340
N001 (  1,  1) [000110] ------------              |  |  \--*  LCL_VAR   ref    V01 arg0         u:2 (last use) $c0
N004 (  7,  6) [000114] -A-X----R---              \--*  ASG       int    $340
N003 (  3,  2) [000113] D------N----                 \--*  LCL_VAR   int    V12 tmp10        d:2 $340

***** BB04, stmt 10
     ( 14, 10) [000032] ------------              *  STMT      void  (IL 0x000...  ???)
N004 (  3,  2) [000172] -------N----              |     /--*  LCL_VAR   int    V12 tmp10        u:2 (last use) $340
N006 (  7,  5) [000173] -A------R---              |  /--*  ASG       int    $340
N005 (  3,  2) [000171] D------N----              |  |  \--*  LCL_VAR   int    V10 tmp8         d:3 $340
N007 ( 14, 10) [000174] -A----------              \--*  COMMA     void   $340
N001 (  3,  2) [000169] -------N----                 |  /--*  LCL_VAR   byref  V11 tmp9         u:2 (last use) $2c0
N003 (  7,  5) [000170] -A------R---                 \--*  ASG       byref  $2c0
N002 (  3,  2) [000168] D------N----                    \--*  LCL_VAR   byref  V09 tmp7         d:3 $2c0
```
None of the copies are removed due to the way copy propagation uses liveness. When a use of a variable is encountered CopyProp looks through the SSA stacks for another variable definition that has the same VN and is live. None of the variables in this IR are live when they are needed.

**Comment by AndyAyersMS** [comment]
Seems like the copy prop code could use dominance instead of liveness for the correctness part of this check. 

There is probably also a perf heuristic embedded in here that might need to be teased out separately.

**Comment by mikedn** [comment]
> There is probably also a perf heuristic embedded in here that might need to be teased out separately.

Yes, when processing an use CopyProp has no way to know that removing the old use would result in the removal of the copy. Without the liveness check it risks extending the live range of a variable without shortening (or completely eliminating) the live range of another variable. Already tried, got 20kbytes of regressions or so.

Maybe it should pay attention to special cases such as variables that have a single definition or definitions that have a single use.

**Comment by mikedn** [comment]
> Seems like the copy prop code could use dominance instead of liveness for the correctness part of this check.

Hmm, something cheaper perhaps: when visiting a block clear out the stacks for variables that aren't live anymore. It's probably more conservative than it needs to be but at least it should not block local copy prop like the current approach does.

**Comment by mikedn** [comment]
@AndyAyersMS Are you still looking into this? I was already looking at fixing other 2 copyprop issues before this came up so I thought it's best to ask first before doing redundant/conflicting work.

**Comment by AndyAyersMS** [comment]
No, not looking at this at present.

**Comment by mikedn** [comment]
> Maybe it should pay attention to special cases such as variables that have a single definition or definitions that have a single use.

I tried something like this. If it's a local definition and it has only one use then ignore liveness. This results in a pretty good improvement but there are also some regressions:
```
Total bytes of diff: -6410 (-0.03% of base)
    diff is an improvement.
Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes
Top file regressions by size (bytes):
         150 : System.Private.Xml.dasm (0.01% of base)
         117 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm (0.00% of base)
          76 : System.Numerics.Vectors.dasm (0.14% of base)
          56 : System.Runtime.Serialization.Formatters.dasm (0.06% of base)
          55 : System.Linq.Expressions.dasm (0.01% of base)
Top file improvements by size (bytes):
       -1834 : System.Private.CoreLib.dasm (-0.05% of base)
       -1504 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.06% of base)
       -1049 : Microsoft.CodeAnalysis.CSharp.dasm (-0.05% of base)
        -371 : Microsoft.CodeAnalysis.dasm (-0.05% of base)
        -335 : System.Net.Http.dasm (-0.10% of base)
74 total files with size differences (49 improved, 25 regressed), 56 unchanged.
Top method regessions by size (bytes):
         312 : System.Linq.Parallel.dasm - QueryOpeningEnumerator`1:OpenQuery():this (26 methods)
         306 : System.Private.Xml.dasm - HtmlElementProps:CreatePropsTable():ref
         145 : System.Private.CoreLib.dasm - DefaultBinder:BindToMethod(int,ref,byref,ref,ref,ref,byref):ref:this
          87 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - ThreadTimeStackComputer:GenerateThreadTimeStacks(ref,ref):this
          57 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - ThreadTimeStackComputer:GetAspNetFromProcessFrameThroughThreadFrameStack(struct,ref,ref):int:this
Top method improvements by size (bytes):
        -239 : System.Private.CoreLib.dasm - OpCodes:.cctor()
        -213 : System.Text.RegularExpressions.dasm - RegexCharClass:.cctor()
        -199 : Microsoft.CodeAnalysis.VisualBasic.dasm - VisualBasicCompilationOptions:.ctor(ref):this
        -140 : Microsoft.VisualBasic.dasm - OverloadResolution:MoreSpecificProcedure(ref,ref,ref,ref,int,byref,bool):ref
        -135 : System.Linq.Parallel.dasm - UnaryQueryOperatorResults:GivePartitionedStream(ref):this (45 methods)
2067 total methods with size differences (1396 improved, 671 regressed), 139090 unchanged.
```
The main problem with doing this is that we need to compute the SSA use counts. That's trivial but then nothing in the JIT will update the counts. And despite what a CopyProp code comment says, CopyProp does not run immediately after SSA and VN, there's also EarlyProp and loop hoisting. Granted, since the use count is not used for correctness purposes this might just work. But it feels like a hack.

**Comment by mikedn** [comment]
Different version that instead of checking that the use count of a SSA def is 1 it only checks that the ref count of the variable is 2:
```
Total bytes of diff: -11509 (-0.05% of base)
    diff is an improvement.
Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes
Top file regressions by size (bytes):
         215 : System.Private.Xml.dasm (0.01% of base)
          97 : System.Numerics.Vectors.dasm (0.17% of base)
          55 : System.Runtime.Serialization.Formatters.dasm (0.06% of base)
          50 : System.Linq.Expressions.dasm (0.01% of base)
          24 : Microsoft.VisualBasic.dasm (0.01% of base)
Top file improvements by size (bytes):
       -5475 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm (-0.21% of base)
       -1698 : System.Private.CoreLib.dasm (-0.05% of base)
       -1438 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.06% of base)
       -1033 : Microsoft.CodeAnalysis.CSharp.dasm (-0.05% of base)
        -384 : Microsoft.CodeAnalysis.dasm (-0.05% of base)
69 total files with size differences (47 improved, 22 regressed), 61 unchanged.
Top method regessions by size (bytes):
         312 : System.Linq.Parallel.dasm - QueryOpeningEnumerator`1:OpenQuery():this (26 methods)
         306 : System.Private.Xml.dasm - HtmlElementProps:CreatePropsTable():ref
         145 : System.Private.CoreLib.dasm - DefaultBinder:BindToMethod(int,ref,byref,ref,ref,ref,byref):ref:this
          87 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - ThreadTimeStackComputer:GenerateThreadTimeStacks(ref,ref):this
          57 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - ThreadTimeStackComputer:GetAspNetFromProcessFrameThroughThreadFrameStack(struct,ref,ref):int:this
Top method improvements by size (bytes):
        -239 : System.Private.CoreLib.dasm - OpCodes:.cctor()
        -213 : System.Text.RegularExpressions.dasm - RegexCharClass:.cctor()
        -199 : Microsoft.CodeAnalysis.VisualBasic.dasm - VisualBasicCompilationOptions:.ctor(ref):this
        -135 : System.Linq.Parallel.dasm - UnaryQueryOperatorResults:GivePartitionedStream(ref):this (45 methods)
        -130 : System.Private.Xml.dasm - XsltLoader:.ctor():this
2667 total methods with size differences (2062 improved, 605 regressed), 138490 unchanged.
```
Hmm, simpler and better. There are still some regressions. Though copy propagation in its current version has regressions as well.

**Comment by mikedn** [comment]
Another example that shows the funny results CopyProp's heuristic can produce:
```C#
struct Foo {
    public int x;
    public int y;
    public int z;
    public int w;
}
static void Test(int x, ref Foo f) {
    Foo f2 = new Foo {
        x = x,
        y = x,
        z = x,
        w = x
    };
    f = f2;
}
```
generates:
```asm
G_M55886_IG01:
       4883EC18             sub      rsp, 24
       C5F877               vzeroupper
       33C0                 xor      rax, rax
       4889442408           mov      qword ptr [rsp+08H], rax
       4889442410           mov      qword ptr [rsp+10H], rax
G_M55886_IG02:
       8BC1                 mov      eax, ecx
       8BC8                 mov      ecx, eax
       448BC0               mov      r8d, eax
       4C8D4C2408           lea      r9, bword ptr [rsp+08H]
       418909               mov      dword ptr [r9], ecx
       45894104             mov      dword ptr [r9+4], r8d
       41894108             mov      dword ptr [r9+8], eax
       4189410C             mov      dword ptr [r9+12], eax
       C4E17A6F442408       vmovdqu  xmm0, qword ptr [rsp+08H]
       C4E17A7F02           vmovdqu  qword ptr [rdx], xmm0
G_M55886_IG03:
       4883C418             add      rsp, 24
       C3                   ret
```
Basically a total mess. The `x` argument got copied 3 times into 3 separate registers and then all are used to initialize the `x`, `y`, `z` and `w` fields. And this still happens even if you get rid of the intermediary `f2` variable:
```asm
       8BC1                 mov      eax, ecx
       8BC8                 mov      ecx, eax
       8902                 mov      dword ptr [rdx], eax
       894204               mov      dword ptr [rdx+4], eax
       894A08               mov      dword ptr [rdx+8], ecx
       894A0C               mov      dword ptr [rdx+12], ecx
```
No idea what can be done in this case. The ref count solution I mentioned above works only when
there's a single chain of copies of a variable. In this case we have parallel copies and the processing order turns out to be unfortunate.

**Comment by AndyAyersMS** [comment]
Probably worth looking into what the preferencing heuristics do to influence allocation in these cases. Give me a day or two.

**Comment by mikedn** [comment]
In the meantime I'll see what can be done about "parallel" copies. CopyProp should probably pick one of the 4 variables and stick with it instead of picking a different variable every time. But which variable? I'd guess the one with the closest definition would be a good start.

**Comment by mikedn** [comment]
Fixed by #19429 

```asm
G_M55890_IG01:
       0F1F440000           nop
G_M55890_IG02:
       4885D2               test     rdx, rdx
       7507                 jne      SHORT G_M55890_IG03
       33C0                 xor      rax, rax
       4533C0               xor      r8d, r8d
       EB08                 jmp      SHORT G_M55890_IG04
G_M55890_IG03:
       488D4210             lea      rax, bword ptr [rdx+16]
       448B4208             mov      r8d, dword ptr [rdx+8]
G_M55890_IG04:
       488901               mov      bword ptr [rcx], rax
       44894108             mov      dword ptr [rcx+8], r8d
       488BC1               mov      rax, rcx
G_M55890_IG05:
       C3                   ret
```
